graviton/dryrun.py

```diff
@@ -0,0 +1,919 @@
+import pytest
+from typing import List, Union
+
+from algosdk.constants import payment_txn, appcall_txn
+from algosdk.future import transaction
+from algosdk.encoding import encode_address, msgpack_encode
+from algosdk.v2client.models import (
+    DryrunRequest,
+    DryrunSource,
+    Application,
+    ApplicationParams,
+    ApplicationStateSchema,
+    Account,
+    TealKeyValue,
+)
+
+
+ZERO_ADDRESS = encode_address(bytes(32))
+PRINTABLE = frozenset(string.printable)
+
+
+@dataclass
+class LSig:
+    """Logic Sig program parameters"""
+
+    args: List[bytes] = None
+
+
+@dataclass
+class App:
+    """Application program parameters"""
+
+    creator: str = ZERO_ADDRESS
+    round: int = None
+    app_idx: int = 0
+    on_complete: int = 0
+    args: List[bytes] = None
+    accounts: List[Union[str, Account]] = None
+    global_state: List[TealKeyValue] = None
+
+
+#### LIGHTWEIGHT ASSERTIONS FOR RE-USE ####
+def _msg_if(msg):
+    return "" if msg is None else f": {msg}"
+
+
+def _fail(msg):
+    assert False, msg
+
+
+def _assert_in(status, msgs, msg=None, enforce=True):
+    ok = status in msgs
+    result = None
+    if not ok:
+        result = f"{status} should be in {msgs}" + _msg_if(msg)
+        if enforce:
+            assert status in msgs, result
+
+    return ok, result
+
+
+def assert_pass(txn_index, msg, txns_res):
+    assert_status("PASS", txn_index, msg, txns_res)
+
+
+def assert_reject(txn_index, msg, txns_res):
+    assert_status("REJECT", txn_index, msg, txns_res)
+
+
+def assert_status(status, txn_index, msg, txns_res, enforce=True):
+    if txn_index is not None and (txn_index < 0 or txn_index >= len(txns_res)):
+        _fail(f"txn index {txn_index} is out of range [0, {len(txns_res)})")
+
+    assert_all = True
+    all_msgs = []
+    if status == "REJECT":
+        assert_all = False
+
+    for idx, txn_res in enumerate(txns_res):
+        # skip if txn_index is set
+        if txn_index is not None and idx != txn_index:
+            continue
+
+        msgs = []
+        if (
+            "logic-sig-messages" in txn_res
+            and txn_res["logic-sig-messages"] is not None
+            and len(txn_res["logic-sig-messages"]) > 0
+        ):
+            msgs = txn_res["logic-sig-messages"]
+        elif (
+            "app-call-messages" in txn_res
+            and txn_res["app-call-messages"] is not None
+            and len(txn_res["app-call-messages"]) > 0
+        ):
+            msgs = txn_res["app-call-messages"]
+        else:
+            _fail("no messages from dryrun")
+        if assert_all or idx == txn_index:
+            _assert_in(status, msgs, msg=msg)
+        all_msgs.extend(msgs)
+
+    if not assert_all:
+        return _assert_in(status, all_msgs, msg=msg, enforce=enforce)
+
+    return True, None
+
+
+def assert_error(drr, contains=None, txn_index=None, msg=None, enforce=True):
+    error = DryRunHelper.find_error(drr, txn_index=txn_index)
+    ok = bool(error)
+    result = None
```

Comment:
@tzaffi Minor - Should be safe to remove the definition.  It's redefined/reassigned in the same lexical scope.

```suggestion

```

Reply:
👍